qnv.py

```diff
@@ -41,42 +28,47 @@ def quatMultiply(q1,q2):
 	
 	a = a1*a2 - np.dot(b1,b2)
 	b = a1*b2 + a2*b1 + np.cross(b1,b2)
-
-	#b = b.reshape((1,3))
-
 	q = np.hstack((a,b))
 	q = q/np.linalg.norm(q)
 	return q
 
-def quatRotate(q,x):
+def quatMultiplyunnorm(q1,q2): #returns quaternion product (product is not a unit quaternion)
+
+	a1 = q1[0:1].copy()
+	a2 = q2[0:1].copy()
 	
-	#rotates vecctor x by quaternion q
-	#M = np.array([[q[0]**2 + q[1]**2 - q[2]**2 - q[3]**2,2*]])
+	b1 = (q1[1:4].copy())
+	b2 = (q2[1:4].copy())
 	
-	qi = np.zeros(4)
+	a = a1*a2 - np.dot(b1,b2)
+	b = a1*b2 + a2*b1 + np.cross(b1,b2)
+	q = np.hstack((a,b))
+	return q
 
-	qi[0] = q[0].copy()
-	qi[1:4] = -1*q[1:4].copy()
+def quatRotate(q,x): #rotates vecctor x by quaternion q
+	
+	if np.count_nonzero(x) == 0:
+		return x
+	qi = quatInv(q)
 	y = np.hstack(([0.],x.copy()))
-	y = quatMultiply(q,y)
-	y = quatMultiply(y,qi)
-
+	y = quatMultiplyunnorm(q,y)
+	y = quatMultiplyunnorm(y,qi)
 	x2 = y[1:4]
 	return x2
 
 def quatDer1(q,w): #if w is in body frame, q takes from body to inertial

```

Comment:
change the comment to : w i angular velocity of body wrt inertial frame in body frame. q transforms inertial frame vector to body frame